File: src/main/java/ru/citeck/ecos/predicate/model/AndPredicate.java
package ru.citeck.ecos.predicate.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class AndPredicate extends ComposedPredicate {

    private static final String TYPE = "and";

    @JsonProperty("t")
    String getType() {
        return TYPE;
    }

    public static List<String> getTypes() {
        return Collections.singletonList(TYPE);
    }

    public static AndPredicate of(Predicate... predicates) {
        AndPredicate and = new AndPredicate();
        and.setPredicates(Arrays.asList(predicates));
        return and;
    }

    @Override
    public String toString() {
        return "(" + getPredicates().stream()
                                    .map(Object::toString)
                                    .collect(Collectors.joining(" AND ")) + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        return super.equals(o);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/model/AndPredicate.kt
package ru.citeck.ecos.records2.predicate.model

import com.fasterxml.jackson.annotation.JsonProperty

class AndPredicate : ComposedPredicate() {

    companion object {

        const val TYPE = "and"

        @JvmStatic
        fun getTypes(): List<String> {
            return listOf(TYPE)
        }

        @JvmStatic
        fun of(vararg predicates: Predicate): AndPredicate {
            return of(listOf(*predicates))
        }

        @JvmStatic
        fun of(predicates: List<Predicate>): AndPredicate {
            val and = AndPredicate()
            and.setPredicates(predicates)
            return and
        }
    }

    @JsonProperty("t")
    fun getType(): String {
        return TYPE
    }

    override fun <T : Predicate> copy(): T {
        val copy = AndPredicate()
        copy.setPredicates(getPredicates().map { it.copy<Predicate>() })
        @Suppress("UNCHECKED_CAST")
        return copy as T
    }

    override fun toString(): String {
        return "{\"t\":\"$TYPE\",\"val\":[${getPredicates().joinToString(",") { it.toString() }}]}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || AndPredicate::class.java != other::class.java) {
            return false
        }
        other as AndPredicate
        return getPredicates() == other.getPredicates()
    }

    override fun hashCode(): Int {
        return getPredicates().hashCode()
    }
}
